lib/Demangling/Demangler.cpp

```diff
@@ -2479,11 +2479,49 @@ NodePointer Demangler::popProtocolConformance() {
       return createNode(Node::Kind::OutlinedBridgedMethod, Params);
     }
     case 'u': return createNode(Node::Kind::AsyncFunctionPointer);
+    case 'J': {
+      auto result = createNode(Node::Kind::AutoDiffFunction);
+      auto optionalGenSig = popNode(Node::Kind::DependentGenericSignature);
+      auto original = popNode();
+      addChild(result, original);
+      addChild(result, optionalGenSig);
+      auto kind = demangleAutoDiffFunctionKind();
+      if (!kind) return nullptr;
+      addChild(result, kind);
+      addChild(result, demangleIndexSubset());
+      if (!nextIf('p')) return nullptr;
+      addChild(result, demangleIndexSubset());
+      if (!nextIf('r')) return nullptr;
+      return result;
+    }
     default:
       return nullptr;
   }
 }
 
+NodePointer Demangler::demangleAutoDiffFunctionKind() {
```

Comment:
I prefer to keep `AutoDiffFunctionKind` in order to be able to switch on an enum rather than raw characters in NodePointer, if you don't mind. I've also simplified it such that the enum has `char` type and character codes are assigned in the enum definition, so this is much simpler now. PTAL.